xbmc/guilib/GUIIncludes.cpp

```diff
@@ -295,6 +300,127 @@ void CGUIIncludes::ResolveIncludesForNode(TiXmlElement *node, std::map<INFO::Inf
+      {
+        string paramName(attribute->Name() + sizeof(paramNamespacePrefix) - 1);
+        if (params.find(paramName) == params.end())
+          params[paramName] = attribute->ValueStr();
+      }
+      attribute = attribute->Next();
+    }
+  }
+  return params;
+}
+
+void CGUIIncludes::ResolveParametersForNode(TiXmlElement *node, const map<string, string> &params) const
+{
+  if (!node)
+    return;
+  string newValue;
+  // run through this element's attributes, resolving any parameters
+  TiXmlAttribute *attribute = node->FirstAttribute();
+  while (attribute)
+  {
+    RESOLVE_PARAMS_RESULT result = ResolveParameters(attribute->ValueStr(), newValue, params);
+    if (result == SINGLE_UNDEFINED_PARAM_RESOLVED && strcmp(node->Value(), "include") == 0)
+    {
+      // special case: passing p:name="$PARAM[undefinedParam]" to the nested include
+      // this usually happens when trying to forward a missing parameter from the enclosing include to the nested include
+      // problem: since 'undefinedParam' is not defined, it expands to p:name="" and overrides any default value set for p:name in the nested include
+      // to prevent this, we'll completely remove this parameter from the nested include call so that the default value can be correctly picked up later
+      const char *name = attribute->Name();
+      attribute = attribute->Next();
+      node->RemoveAttribute(name);
+    }
+    else
+    {
+      if (result != NO_PARAMS_FOUND)
+        attribute->SetValue(newValue);
+      attribute = attribute->Next();
+    }
+  }
+  // run through this element's value and children, resolving any parameters
+  if (TiXmlNode *child = node->FirstChild())
+  {
+    if (child->Type() == TiXmlNode::TINYXML_TEXT)
+    {
+      if (ResolveParameters(child->ValueStr(), newValue, params) != NO_PARAMS_FOUND)
+        child->SetValue(newValue);
+    }
+    else if (child->Type() == TiXmlNode::TINYXML_ELEMENT)
+    {
+      do
+      {
+        ResolveParametersForNode(static_cast<TiXmlElement*>(child), params);
+        child = child->NextSiblingElement();
+      }
+      while (child);
+    }
+  }
+}
+
+class ParamReplacer
+{
+  const map<string, string> &m_params;
+  // keep some stats so that we know exactly what's been resolved
+  int m_numTotalParams;
+  int m_numUndefinedParams;
+public:
+  ParamReplacer(const map<string, string> &params) : m_params(params), m_numTotalParams(0), m_numUndefinedParams(0) {}
+  int getNumTotalParams() const { return m_numTotalParams; }
+  int getNumDefinedParams() const { return m_numTotalParams - m_numUndefinedParams; }
+  int getNumUndefinedParams() const { return m_numUndefinedParams; }
+  string operator()(const string &paramName)
+  {
+    m_numTotalParams++;
+    map<string, string>::const_iterator it = m_params.find(paramName);
+    if (it != m_params.end())
+      return it->second;
+    else
+    {
+      m_numUndefinedParams++;
+      return StringUtils::Empty;
```

Comment:
Just use `return "";`. `StringUtils::Empty` is meant to be used if a method expects a `const std::string&` as a return type.